docs/spec/overload.rst

```diff
@@ -99,8 +109,349 @@ variable is not sufficient.
+
+If all possible sets of arguments accepted by an overload are also always
+accepted by an earlier overload, the two overloads are said to "fully overlap".
+In this case, the latter overload will never be used. This condition
+is indicative of a programming error and should be reported by type
+checkers::
+
+  # These overloads fully overlap because the first overload
+  # accepts all arguments accepted by the second overload.
+
+  @overload
+  def func[T](x: T) -> T: ...
+  @overload
+  def func(x: int) -> int: ...
+
+
+Overload call evaluation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+When a type checker evaluates the call of an overloaded function, it
+attempts to "match" the supplied arguments with one or more overloads.
+This section describes the algorithm that type checkers should use
+for overload matching. This algorithm should be applied even in the
+presence of :ref:<overlapping overloads>.
+
+Only the overloads (the ``@overload``-decorated signatures) should be
+considered for matching purposes. The implementation, if provided,
+should be ignored for purposes of overload matching.
+
+
+Step 1: Examine the argument list to determine the number of
+positional and keyword arguments. Use this information to eliminate any
+overload candidates that are not plausible based on their
+input signatures.
+
+- If no candidate overloads remain, generate an error and stop.
+- If only one candidate overload remains, it is the winning match. Evaluate
+  it as if it were a non-overloaded function call and stop.
+- If two or more candidate overloads remain, proceed to step 2.
+
+
+Step 2: Evaluate each remaining overload as a regular (non-overloaded)
+call to determine whether it is compatible with the supplied
+argument list. Unlike step 1, this step considers the types of the parameters
+and arguments. During this step, do not generate any user-visible errors.
+Simply record which of the overloads result in evaluation errors.
+
+- If all overloads result in errors, proceed to step 3.
+- If only one overload evaluates without error, it is the winning match.
+  Evaluate it as if it were a non-overloaded function call and stop.
+- If two or more candidate overloads remain, proceed to step 4.
+
+
+Step 3: If step 2 produces errors for all overloads, perform
+"argument type expansion". Union types can be expanded
+into their constituent subtypes. For example, the type ``int | str`` can
+be expanded into ``int`` and ``str``.
+
+Type expansion should be performed one argument at a time from left to
+right. Each expansion results in sets of effective argument types.
+For example, if there are two arguments whose types evaluate to
+``int | str`` and ``int | bytes``, expanding the first argument type
+results in two sets of argument types: ``(int, int | bytes)`` and
+``(str, int | bytes)``. If type expansion for the second argument is required,
+four sets of argument types are produced: ``(int, int)``, ``(int, bytes)``,
+``(str, int)``, and ``(str, bytes)``.
```

Comment:
I'd like to suggest adding that when a function is an overload implementation then some combinations may be further eliminated based on the overload definitions. This makes it easier to call one overloaded function from the implementation of another overloaded function with compatible type arguments. For example:
```python
from typing import overload

@overload
def take_varg(*arg: int) -> int:
    ...
@overload
def take_varg(*arg: str) -> str:
    ...

@overload
def take_varg2(*arg: int) -> int:
    ...
@overload
def take_varg2(*arg: str) -> str:
    ...
def take_varg2[T: (int, str)](*arg: T) -> T:
    return take_varg(*arg)

@overload
def take_pair(arg: int, arg2: int) -> int:
    ...
@overload
def take_pair(arg: str, arg2: str) -> str:
    ...
@overload
def take_pair2(arg: int, arg2: int) -> int:
    ...
@overload
def take_pair2(arg: str, arg2: str) -> str:
    ...
def take_pair2[T: (int, str)](arg: T, arg2: T) -> T:
    return take_pair(arg, arg2)
```
The above currently type checks in mypy but not pyright. I've used "value-constrained type variables" which based on a [related discussion](https://github.com/microsoft/pyright/issues/9475#issuecomment-2481225231) on pyright "are not well specified in the Python typing spec", and you recommend against using them.